### PR TITLE
Move table representation builder

### LINKED ISSUE #408: `PhotometryData` cannot read files it has written
It turns out that the place where we add code to astropy table so that it can understand the objects we are saving to table metadata is [here in `settings/models.py`](https://github.com/feder-observatory/stellarphot/blob/2.0.0.alpha1/stellarphot/settings/models.py#L131). This code is executed whenever you create an instance of one of our models.

A side effect of that design decision is that `PhotometryData` cannot read a file it has written unless and until an instance of each model is created. It may be that just creating a `PhotometrySettings` object is adequate.

In any event, this code, along with the attached file [photometry.ecsv.zip](https://github.com/user-attachments/files/16551832/photometry.ecsv.zip), is enough to demonstrate the bug (unzip the file first).

```
from stellarphot import PhotometryData
PhotometryData.read("photometry.ecsv")
```

then you get this traceback:

```python-traceback
---------------------------------------------------------------------------
ConstructorError                          Traceback (most recent call last)
File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/meta.py:431](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/meta.py#line=430), in get_header_from_yaml(lines)
    430 try:
--> 431     header = yaml.load(header_yaml, Loader=TableLoader)
    432 except Exception as err:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/__init__.py:81](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/__init__.py#line=80), in load(stream, Loader)
     80 try:
---> 81     return loader.get_single_data()
     82 finally:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:51](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=50), in BaseConstructor.get_single_data(self)
     50 if node is not None:
---> 51     return self.construct_document(node)
     52 return None

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:60](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=59), in BaseConstructor.construct_document(self, node)
     59 for generator in state_generators:
---> 60     for dummy in generator:
     61         pass

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:413](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=412), in SafeConstructor.construct_yaml_map(self, node)
    412 yield data
--> 413 value = self.construct_mapping(node)
    414 data.update(value)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:218](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=217), in SafeConstructor.construct_mapping(self, node, deep)
    217     self.flatten_mapping(node)
--> 218 return super().construct_mapping(node, deep=deep)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:143](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=142), in BaseConstructor.construct_mapping(self, node, deep)
    141     raise ConstructorError("while constructing a mapping", node.start_mark,
    142             "found unhashable key", key_node.start_mark)
--> 143 value = self.construct_object(value_node, deep=deep)
    144 mapping[key] = value

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:100](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=99), in BaseConstructor.construct_object(self, node, deep)
     99 if tag_suffix is None:
--> 100     data = constructor(self, node)
    101 else:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py:427](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/yaml/constructor.py#line=426), in SafeConstructor.construct_undefined(self, node)
    426 def construct_undefined(self, node):
--> 427     raise ConstructorError(None, None,
    428             "could not determine a constructor for the tag %r" % node.tag,
    429             node.start_mark)

ConstructorError: could not determine a constructor for the tag '!Camera'
  in "<unicode string>", line 41, column 13:
        camera: !Camera {model_json_string: '{"n ... 
                ^

The above exception was the direct cause of the following exception:

YamlParseError                            Traceback (most recent call last)
File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ecsv.py:157](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ecsv.py#line=156), in EcsvHeader.get_cols(self, lines)
    156 try:
--> 157     header = meta.get_header_from_yaml(lines)
    158 except meta.YamlParseError:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/meta.py:433](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/meta.py#line=432), in get_header_from_yaml(lines)
    432 except Exception as err:
--> 433     raise YamlParseError() from err
    435 return header

YamlParseError: 

During handling of the above exception, another exception occurred:

InconsistentTableError                    Traceback (most recent call last)
Cell In[2], line 1
----> 1 pd = PhotometryData.read("photometry.ecsv")

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/connect.py:62](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/table/connect.py#line=61), in TableRead.__call__(self, *args, **kwargs)
     59 units = kwargs.pop("units", None)
     60 descriptions = kwargs.pop("descriptions", None)
---> 62 out = self.registry.read(cls, *args, **kwargs)
     64 # For some readers (e.g., ascii.ecsv), the returned `out` class is not
     65 # guaranteed to be the same as the desired output `cls`.  If so,
     66 # try coercing to desired class without copying (io.registry.read
     67 # would normally do a copy).  The normal case here is swapping
     68 # Table <=> QTable.
     69 if cls is not out.__class__:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/registry/core.py:221](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/registry/core.py#line=220), in UnifiedInputRegistry.read(self, cls, format, cache, *args, **kwargs)
    218         kwargs.update({"filename": path})
    220 reader = self.get_reader(format, cls)
--> 221 data = reader(*args, **kwargs)
    223 if not isinstance(data, cls):
    224     # User has read with a subclass where only the parent class is
    225     # registered.  This returns the parent class, so try coercing
    226     # to desired subclass.
    227     try:

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/connect.py:19](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/connect.py#line=18), in io_read(format, filename, **kwargs)
     17     format = re.sub(r"^ascii\.", "", format)
     18     kwargs["format"] = format
---> 19 return read(filename, **kwargs)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py:603](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py#line=602), in deprecated_renamed_argument.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    600             msg += f"\n        Use {alternative} instead."
    601         warnings.warn(msg, warning_type, stacklevel=2)
--> 603 return function(*args, **kwargs)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py:603](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py#line=602), in deprecated_renamed_argument.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    600             msg += f"\n        Use {alternative} instead."
    601         warnings.warn(msg, warning_type, stacklevel=2)
--> 603 return function(*args, **kwargs)

    [... skipping similar frames: deprecated_renamed_argument.<locals>.decorator.<locals>.wrapper at line 603 (2 times)]

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py:603](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/utils/decorators.py#line=602), in deprecated_renamed_argument.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    600             msg += f"\n        Use {alternative} instead."
    601         warnings.warn(msg, warning_type, stacklevel=2)
--> 603 return function(*args, **kwargs)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ui.py:461](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ui.py#line=460), in read(table, guess, **kwargs)
    459     else:
    460         reader = get_reader(**new_kwargs)
--> 461         dat = reader.read(table)
    462         _read_trace.append(
    463             {
    464                 "kwargs": copy.deepcopy(new_kwargs),
   (...)
    467             }
    468         )
    470 # Static analysis (pyright) indicates `dat` might be left undefined, so just
    471 # to be sure define it at the beginning and check here.

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/core.py:1431](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/core.py#line=1430), in BaseReader.read(self, table)
   1428 self.header.update_meta(self.lines, self.meta)
   1430 # Get the table column definitions
-> 1431 self.header.get_cols(self.lines)
   1433 # Make sure columns are valid
   1434 self.header.check_column_names(self.names, self.strict_names, self.guessing)

File [~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ecsv.py:159](http://localhost:8888/lab/tree/~/miniforge3/envs/stelldev-pyd2/lib/python3.11/site-packages/astropy/io/ascii/ecsv.py#line=158), in EcsvHeader.get_cols(self, lines)
    157     header = meta.get_header_from_yaml(lines)
    158 except meta.YamlParseError:
--> 159     raise core.InconsistentTableError("unable to parse yaml in meta header")
    161 if "meta" in header:
    162     self.table_meta = header["meta"]

InconsistentTableError: unable to parse yaml in meta header
```

### PR DESCRIPTION
**Pull request by mwcraig**
Fixes #408.

The root cause of #408 was that the mechanism for allowing an astropy `Table` to read/write our custom models was built into the model classes and not run unless one instantiated each of the models.  As a result, if you tried reading a `PhotometryData` table before you instantiated a camera and an observatory then the read would fail.

This PR fixes the issue by ensuring that the code to add the table representations is run when stellarphot is imported. 

This issue is hard to test for because if any code anywhere in the test suite makes a `Camera` and an `Observatory` hen the test passes. To ensure that we are testing for this failure I added a special tox case for it and added a workflow that runs that test.

### CODE REVIEW

**Comment by mwcraig** [comment]
Marked as a draft to make sure we get #419 in first

File: stellarphot/tests/test_core_minimal_imports.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from astropy.utils.data import get_pkg_data_filename
+
+from stellarphot.core import PhotometryData
+
+
+# Why is this in a separate file?
+#
+# The bug we are trying to reproduce happens only no objects from
```
Comment (JuanCab) [inline stellarphot/tests/test_core_minimal_imports.py:8]: Incomplete phrase:
```suggestion
# The bug we are trying to reproduce happens only when no objects from
```

File: .github/workflows/build.yml
Comment (JuanCab) [inline .github/workflows/build.yml:1]: I'm not fully conversant in or even have any experience at all withtox. So I did look over the tox stuff and understand the idea, although I can't really be relied on to catch any bugs in it. :(

File: stellarphot/tests/test_core_minimal_imports.py
Comment (JuanCab) [inline stellarphot/tests/test_core_minimal_imports.py:1]: Nice solution to the testing problem.

**Review by JuanCab — APPROVED** [review]
Interesting fix.  I can't claim to understand the `tox` aspects of this, but the logic seemed sound.

File: .github/workflows/build.yml
Comment (mwcraig) [reply .github/workflows/build.yml:1]: I was marginally surprised the tox changes worked 🤣

**Comment by mwcraig** [comment]
@JuanCab  -- is this OK to merge once tests pass?

**Comment by JuanCab** [comment]
> @JuanCab -- is this OK to merge once tests pass?

Yep

**Pull request merged**